File: implementations/flutter/design_patterns_flutter/lib/patterns/behavioral/template/concrete_signup.dart
import 'abstract_class.dart';

class SignupFlow extends FormTemplate {
  @override
  Map<String, dynamic> validate(Map<String, dynamic> data) {
    if (data['email'] == null || (data['email'] as String).trim().isEmpty) {
      throw Exception('Email is required');
    }
    if (data['password'] == null || (data['password'] as String).isEmpty) {
      throw Exception('Password is required');
    }
    return {
      'email': (data['email'] as String).trim().toLowerCase(),
      'password': data['password'],
    };
  }

  @override
  Future<dynamic> send(Map<String, dynamic> cleanData) async {
    // simulate network call
    print('[Signup] sending to server: $cleanData');
    return {
      'userId': 'user_' + DateTime.now().millisecondsSinceEpoch.toString(),
      'email': cleanData['email'],
    };
  }

  @override
  Future<void> afterSuccess(result) async {
    // pretend we send analytics
    print('[Signup] analytics event for user ${result['userId']}');
  }

  @override
  String successMessage(result) {
    return 'Welcome, ${result['email']}! Your account is ready.';
  }
}

File: implementations/react-native/design-patterns-rn/src/patterns/behavioral/template/concrete-signup.ts
import { FormTemplate } from './abstract-class';

export class SignupFlow extends FormTemplate {
  protected validate(data: Record<string, any>) {
    if (!data.email) throw new Error('Email is required');
    if (!data.password) throw new Error('Password is required');
    return {
      email: String(data.email).trim().toLowerCase(),
      password: data.password,
    };
  }

  protected async send(cleanData: Record<string, any>) {
    console.log('[Signup] sending to server:', cleanData);
    return {
      userId: 'user_' + Math.random().toString(36).slice(2),
      email: cleanData.email,
    };
  }

  protected async afterSuccess(result: any) {
    console.log('[Signup] analytics event for user', result.userId);
  }

  protected successMessage(result: any) {
    return `Welcome, ${result.email}! Your account is ready.`;
  }
}
